docs/dynamic-programming/kadane-algorithm.mdx

```diff
@@ -0,0 +1,222 @@
+---
+title: Kadane's Algorithm
+slug: /dp/kadane-algorithm
+---
+
+## Introduction
+
+Kadane's Algorithm gives us the maximum sum out of all possible contiguous subarray in a one-dimensional array. It is a Dynamic Programming based algorithm because it uses the previously computed data for further computations. It is the most optimal solution for finding the maximum sum contiguous subarray. It has many applications like genomic sequence analysis and computer vision.
+
+## Explanation 
+
+Contiguous subarray refers to the continuous part of an array. Contiguous subarrays of `{p,q,r}` are `{p}`, `{q}`, `{r}`, `{p,q}`, `{q,r}`, `{p,q,r}`. This algorithm iterates over all positive-sum contiguous subarrays and keep a track of the maximum sum contiguous subarray. If the sum of the current subarray on which we are iterating is greater than the maximum sum, then we will replace the maximum sum with the current sum.
+
+Let's have one example: 
+<img src="https://i.imgur.com/DKoovMD.png" width="365" height="63" align="center"/><br/><br/>
+
+```
+arr = -4 -1 2 3 5 6 1 -4 -6 
+sum_till_now = arr[0] = -4        
+max_sum = arr[0] = -4
+ 
+Iteration 1:
+if (sum_till_now < 0) sum_till_now = 0
+sum_till_now = sum_till_now + arr[1] = -1
+max_sum = max(max_sum, sum_till_now) = -1
+ 
+Iteration 2:
+if (sum_till_now < 0) sum_till_now = 0
+sum_till_now = sum_till_now + arr[2] = 2
+max_sum = max(max_sum, sum_till_now) = 2
+ 
+Iteration 3:
+if (sum_till_now < 0) sum_till_now = 0
+sum_till_now = sum_till_now + arr[3] = 5
+max_sum = max(max_sum, sum_till_now) = 5
+ 
+Iteration 4:
+if (sum_till_now < 0) sum_till_now = 0
+sum_till_now = sum_till_now + arr[4] = 10
+max_sum = max(max_sum, sum_till_now) = 10
+
+Iteration 5:
+if (sum_till_now < 0) sum_till_now = 0
+sum_till_now = sum_till_now + arr[5] = 16
+max_sum = max(max_sum, sum_till_now) = 16
+ 
+Iteration 6:
+if (sum_till_now < 0) sum_till_now = 0
+sum_till_now = sum_till_now + arr[6] = 17
+max_sum = max(max_sum, sum_till_now) = 17
+ 
+Iteration 7:
+if (sum_till_now < 0) sum_till_now = 0
+sum_till_now = sum_till_now + arr[7] = 13
+max_sum = max(max_sum, sum_till_now) = 17
+ 
+Iteration 8:
+if (sum_till_now < 0) sum_till_now = 0
+sum_till_now = sum_till_now + arr[8] = 7
+max_sum = max(max_sum, sum_till_now) = 17
+```
+
+## Algorithm
+
+- Initialize `max_sum` and `sum_till_now` to `arr[0]`.
+
+- Initialize `i to 1` begin a While loop till `i != n`
+  - If `(sum_till_now < 0)` set it to `0`
+  - `sum_till_now += arr[i]`
+  - `max_sum` = `maximum(max_sum, sum_till_now)`
+  - `i += 1`.
+- Return `max_sum`.
```

Comment:
Use text for writing the algorithm. You can add a separate pseudo-code section but it isn't necessary.

Reply:
done